### PR TITLE
quant fused conv-bn: calculate conv_w rescale without grad

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49981 quant fused conv-bn: calculate conv_w rescale without grad**

Summary:

While analyzing the long standing issue with gradient explosions
when QAT is trained from scratch, I noticed that we are backpropogating
to `bn.weight` when calculating the scaling factor for a conv-bn's conv
weights.  We should discuss in person, but I'm guessing that this was
not intended. Please see https://arxiv.org/pdf/1806.08342.pdf Figure 9
for details.

In an unrealistic experiment, this change improves the stability of
the conv-bn block, although it does not eliminate it completely.  Setup:
* Create torchvision's quantizeable MobileNetV2 with random weights
* fuse it
* set LR to 1.0 (smaller values also lead to a repro, 1.0 is convenient for a fast repro, although not very realistic).
* measure the # of iterations from start of training to seeing the first NaNs
* the number ^ is ~20 iterations without this PR, and increases to ~100 iterations with this PR
* (more analysis is FB-only for now, TODO write it up)

Note: I've traced the further instability to the interplay between bn's
running variance and the conv weight scaling, we can discuss a separate
recommendation for that outside of this PR.

Note: only the `with torch.no_grad()` line is for commit,
the rest is debugging info which is convenient to have during analysis
and will be removed before landing.

Test Plan:

See above.
TODO: experiments on a more realistic setting

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by vkuzo** [comment]
abandoning for now

**Comment by fxmarty** [comment]
@vkuzo Do you have an indication of where the fusing for Figure 9 in https://arxiv.org/pdf/1806.08342.pdf is implemented in PyTorch? I am trying to reproduce it but it is unclear from the figure how the backward graph should look like.

Alternatively, figure C.7 and C.8 in https://arxiv.org/pdf/1712.05877.pdf suggest that you should `.detach()` BN's weight, bias, and the conv-layer weight at some point. Is it the case?

**Comment by fxmarty** [comment]
Oh my bad it is here https://github.com/pytorch/pytorch/blob/master/torch/nn/intrinsic/qat/modules/conv_fused.py#L258 .
